lib/fs.js

```diff
@@ -2741,6 +2752,67 @@ function mkdtempSync(prefix, options) {
   return result;
 }
 
+/**
+ * Asynchronously copies `src` to `dest`. `src` can be a file, directory, or
+ * symlink. The contents of directories will be copied recursively.
+ * @param {string | Buffer | URL} src
+ * @param {string | Buffer | URL} dest
+ * @param {Object} [options]
+ * @param {() => any} callback
+ * @returns {void}
+ */
+function copy(src, dest, options, callback) {
+  if (typeof options === 'function' && !callback) {
+    callback = options;
+    options = {};
+  } else if (typeof options === 'function') {
+    options = { filter: options };
+  }
+  callback = makeCallback(callback);
+  options = options || {};
+  options = validateCopyOptions(options);
+  if (isBuffer(src)) {
+    src = src.toString();
+  }
+  if (isBuffer(dest)) {
+    dest = dest.toString();
+  }
```

Comment:
I don't think this is necessary, and at least it shouldn't come before `getValidatedPath`: https://github.com/nodejs/node/blob/8886b63cf66c29d453fdc1ece2e489dace97ae9d/lib/internal/fs/utils.js#L667-L683

Reply:
@aduh95 I pulled the logic that turns the buffer into a string into `copy`/`copySync`, which I think is a reasonable step for now and cleans up the logic in `fs.js`.

Reply:
Then maybe we should remove the support? That would make this function the weird one out, but it seems converting silently to strings could cause some unexpected behavior: https://github.com/nodejs/node/blob/6cd12be34765d9d618d063059b80ed11ee99bac5/doc/api/fs.md#L6578-L6582

FWIW, all the other FS methods also accepts `Uint8Array` instances, not just `Buffer` ones (even though it's not documented); the `isBuffer` check is probably not the right one.

Reply:
I'd rather drop support for `Buffer`, and allow for URL and `string`, this functionality could be added later, but we'd need to refactor the path handling in `copy` to support it.

Reply:
Dropping the support for `Buffer` is.... difficult. It's one of those things where probably less than 1% of Node.js users would ever actually use `Buffer` here but it was added specifically because of a case I ran into where a linux file path included three distinct character encodings in a single path and things just... broke. I'm not going to block on this, but it is definitely something to keep in mind.

Reply:
@jasnell @aduh95 I don't believe the problem is the file operations, as it's using other API methods that accept a `Buffer`, the problem is I don't believe the `path` helpers accept a buffer.

Perhaps the follow up work is that it would be good if they did?

Reply:
I think we should remove the `Buffer` mention in the documentation for `copy[Sync]` methods, and give them a `Stability 1: Experimental` status. One condition to upgrade to stable status could be to have a proper `Buffer` support, but it doesn't have to block this PR IMHO.

Reply:
@jasnell @aduh95 my concern is that I'm not sure what proper Buffer support would look like, to handle the case of a directory structure with different file encodings, I believe we would need to set `encoding`, `buffer` always on the `readdir` operation. At which point, what would the logic be for joining paths?

I suppose the question is, is there any safe way to convert the buffer to a string?

Reply:
Reading through `utils.js`, it does look like we have some helpers that can handle Buffers:

```js
function join(path, name) {
  if ((typeof path === 'string' || isUint8Array(path)) &&
      name === undefined) {
    return path;
  }

  if (typeof path === 'string' && isUint8Array(name)) {
    const pathBuffer = Buffer.from(pathModule.join(path, pathModule.sep));
    return Buffer.concat([pathBuffer, name]);
  }

  if (typeof path === 'string' && typeof name === 'string') {
    return pathModule.join(path, name);
  }

  if (isUint8Array(path) && isUint8Array(name)) {
    return Buffer.concat([path, bufferSep, name]);
  }

  throw new ERR_INVALID_ARG_TYPE(
    'path', ['string', 'Buffer'], path);
}
```

Reply:
A bit more digging, the real problem I believe is `dirname`. If I understand correctly, this would be difficult to implement because the separator character `\` or `/` could actually just be part of a multi-byte unicode character?

Reply:
The other concern, along with the `dirname` method, is the `filter` helper. As follow on work, would we be okay with this solution:

1. implement a `dirname` that attempts to be smart about UTF-8 escape sequences (_perhaps it could walk from the front of the buffer to the back, and would ignore separator characters that appear to be part of an escape sequence).
2. call `Buffer.toString()` on the value passed to the filter function, I understand that this won't be perfect, but it's potentially _ good enough_? it seems like it would put a lot of burden on someone writing the filter function to handle a Buffer, no?

Reply:
I think we should pass the buffer directly to the `filter` function. If the user want want to support `Buffer` as paths, they should implement support for `Buffer` in the `filter` function. We should add an option `encoding` so if they want to opt-out of `Buffer` support, they can.
For `dirname`, we might need to add support for `Uint8Array` in `dirname` itself. A discussion for another PR I think.